jcstress-samples/src/main/java/org/openjdk/jcstress/samples/concurreny/mutex/Mutex_01_NoAlgorithm.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2014, 2015, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.jcstress.samples.concurreny.mutex;
+
+import org.openjdk.jcstress.annotations.Actor;
+import org.openjdk.jcstress.annotations.JCStressTest;
+import org.openjdk.jcstress.annotations.Outcome;
+import org.openjdk.jcstress.annotations.State;
+import org.openjdk.jcstress.infra.results.II_Result;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.openjdk.jcstress.annotations.Expect.ACCEPTABLE;
+
+
+public class Mutex_01_NoAlgorithm {
+
+    /*
+        How to run this test:
+            $ java -jar jcstress-samples/target/jcstress.jar -t Mutex_01_NoAlgorithm
+     */
+    /**
+     * This sample demonstrates you how you can introduce a critical section to check algorithms
+     * which ensure only one actor at most can enter the critical section.
+     */
+    @JCStressTest
+    @Outcome(id = {"1, 1", "1, 2", "2, 1", "2, 2"}, expect = ACCEPTABLE, desc = "Both actors can enter the critical section one by one but must not")
+    @State
+    public static class NoAlgorithm {
+        private final AtomicBoolean isInCriticalSection = new AtomicBoolean();
+
+        @Actor
+        public void actor1(II_Result r) {
+            if (isInCriticalSection.compareAndSet(false, true)) {
+                r.r1 = 1;
+                isInCriticalSection.set(false);
+            } else {
+                r.r1 = 2;
+            }
```

Comment:
This kind of code introduces its own synchronization, so it might hide some outcomes. I believe you want to demonstrate the "already taken" detection code works? If so, I suggest doing it this way:

```
boolean taken1, taken2;

@Actor
public void actor1(ZZ_Result r) {
    taken1 = true;
    r.r1 = taken2;
    taken1 = false;
}

@Actor
public void actor2(ZZ_Result r) {
    taken2 = true;
    r.r2 = taken1;
    taken2 = false;
}
```

Reply:
Yes, I tried to find a way to find out whether both actors have entered the critical section at the same time.
Your solution is brilliant! At least one of both actors will always witness in the result that both actors have entered the critical section if they did. However, I think the booleans taken1 and taken2 must be volatile because 
a) they run in different threads with their local caches and might not notice changes of the booleans from the other thread
b) volatile forces the order of those 3 statements. Otherwise, the JVM is allowed to change the order in which they are executed.
Am I right or wrong with volatile here? 
If I am wrong, I will remove the volatile again, of course.

Reply:
`taken1` and `taken2` should not be `volatile` they should be plain fields.
If you make them volatile, you forbid compiler/CPU from making optimizations which might unveil the flaw in the tested algorithms.

For instance, if you add `SynchonizedAlgorithm` that works like `synchronized(String.class) {...}`, then the test must pass even with non-volatile `taken1` and `taken2`. Dekker and other algorithms should work exactly as `synchronized(...)`, so volatile is not needed.